### PR TITLE
[Merged by Bors] - Check proposer index during block production

### LINKED ISSUE #2612: Check proposer index when producing a block
## Description

It came to my attention that it's possible for the VC to publish a block with an invalid signature. Here's the flow:

1. VC gets proposer duties from the BN
1. BN re-orgs
1. VC requests block from the BN, signs it and returns it.
1. BN publishes it, then verifies it and realises that the block is signed by the wrong validator index.

## Solutions

There are two checks we can implement to prevent this:

1. Verify the `randao_reveal` in the BN when [producing a block](https://github.com/sigp/lighthouse/blob/f9bba92db3468321b28ddd9010e26b359f88bafe/beacon_node/beacon_chain/src/beacon_chain.rs#L2604-L2623). Presently we verify the block [without checking signatures](https://github.com/sigp/lighthouse/blob/f9bba92db3468321b28ddd9010e26b359f88bafe/beacon_node/beacon_chain/src/beacon_chain.rs#L2782) to speed up block publishing times. If we checked the `randao_reveal` we'd notice that it was signed by the wrong proposer.
1. In the VC, after we [get the block from the BN](https://github.com/sigp/lighthouse/blob/f9bba92db3468321b28ddd9010e26b359f88bafe/validator_client/src/block_service.rs#L269-L273), check that the proposer index matches our request.

(2) is trivial to implement in terms of runtime cost. (1) is less trivial, but that signature verification should take only a few ms so I think it's worth it.

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Resolves #2612 

## Proposed Changes

Implements both the checks mentioned in the original issue. 
1. Verifies the `randao_reveal` in the beacon node
2. Cross checks the proposer index after getting back the block from the beacon node.

## Additional info
The block production time increases by ~10x because of the signature verification on the beacon node (based on the `beacon_block_production_process_seconds` metric) when running on a local testnet.

### CODE REVIEW

**Comment by michaelsproul** [comment]
> The block production time increases by ~10x because of the signature verification on the beacon node (based on the beacon_block_production_process_seconds metric) when running on a local testnet.

I think we should deploy to Prater to see what the impact is there. What time are we talking in absolute terms? Hopefully only a few milliseconds?

Block production on Prater takes 380ms, so the relative impact is probably very small.

**Comment by pawanjay176** [comment]
Running this on prater,  `per_block_processing` has jumped to take ~10 ms more than before.

Screenshot from 2021-10-26 15-08-25

Total block production times seems to be almost similar

Screenshot from 2021-10-26 15-12-50

(Binary was updated at around 18:50 25/10 in the metrics)

**Comment by michaelsproul** [comment]
Yeah, agree. The impact isn't even visible on the mean/median graph, so I think we're good to go! Will wait for @paulhauner's blessing before merging

File: consensus/state_processing/src/per_block_processing.rs
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing.rs:129]: Ooh, I think this should also be `True` in the case of `BlockSignatureStrategy::VerifyIndividual`. We could use `verify_signatures` in the `else` case to accomplish this (and guard against future additions).

**Review by michaelsproul — CHANGES_REQUESTED** [review]

File: consensus/state_processing/src/per_block_processing/tests.rs
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing/tests.rs]: This should probably stay as `VerifyIndividual` as I think this test failure is pointing to the issue I raised above

File: consensus/state_processing/src/per_block_processing.rs
Comment (pawanjay176) [reply consensus/state_processing/src/per_block_processing.rs:129]: Thanks for catching this. Fixed it.

**Review by michaelsproul — APPROVED** [review]
Nice! LGTM

**Review by paulhauner — APPROVED** [review]
Looks good!

**Pull request closed without merging**
